reana_job_controller/job_monitor.py

```diff
@@ -22,80 +22,97 @@
 from reana_job_controller import config
 from reana_job_controller.htcondorcern_job_manager import \
     HTCondorJobManagerCERN
+from reana_job_controller.job_db import JOB_DB
 from reana_job_controller.kubernetes_job_manager import KubernetesJobManager
 
 
-def watch_jobs_kubernetes(job_db):
-    """Open stream connection to k8s apiserver to watch all jobs status.
+class KubernetesJobsMonitor():
+    """Kubernetes jobs monitor."""
 
-    :param job_db: Dictionary which contains all current jobs.
-    """
-    while True:
-        logging.debug('Starting a new stream request to watch Jobs')
-        try:
-            w = watch.Watch()
-            for event in w.stream(
-                    current_k8s_batchv1_api_client.list_job_for_all_namespaces
-            ):
-                logging.info(
-                    'New Job event received: {0}'.format(event['type']))
-                job = event['object']
+    def __new__(cls):
+        """Check if instance already exists."""
+        if not hasattr(cls, 'instance'):
+            cls.instance = super(KubernetesJobsMonitor, cls).__new__(cls)
+        return cls.instance
 
-                # Taking note of the remaining jobs since deletion might not
-                # happen straight away.
-                remaining_jobs = dict()
-                for job_id, job_dict in job_db.items():
-                    if not job_db[job_id]['deleted']:
-                        remaining_jobs[job_dict['backend_job_id']] = job_id
-                if (not job_db.get(remaining_jobs.get(job.metadata.name)) or
-                        job.metadata.name not in remaining_jobs):
-                    # Ignore jobs not created by this specific instance
-                    # or already deleted jobs.
-                    continue
-                job_id = remaining_jobs[job.metadata.name]
-                kubernetes_job_id = job.metadata.name
-                if job.status.succeeded:
-                    logging.info(
-                        'Job job_id: {}, kubernetes_job_id: {}'
-                        ' succeeded.'.format(job_id, kubernetes_job_id)
-                    )
-                    job_db[job_id]['status'] = 'succeeded'
-                elif (job.status.failed and
-                      job.status.failed >= config.MAX_JOB_RESTARTS):
+    def __init__(self):
+        """Initialize Kubernetes job monitor thread."""
+        job_event_reader_thread = threading.Thread(
```

Comment:
What do you think about storing the thread as an instance attribute just in case we have to perform any operation later?